src/drivers/gnss/septentrio/septentrio.cpp

```diff
@@ -0,0 +1,1702 @@
+	size_t buffer_offset = 0;   // The offset into the string where the next data should be read to.
+	hrt_abstime timeout_time = hrt_absolute_time() + 5 * 1000 * k_receiver_ack_timeout;
+	bool response_detected = false;
+
+	// Receiver prints prompt after a message.
+	if (!send_message(k_command_ping)) {
+		return PX4_ERROR;
+	}
+
+	do {
+		// Read at most the amount of available bytes in the buffer after the current offset, -1 because we need '\0' at the end for a valid string.
+		int read_result = read(reinterpret_cast<uint8_t *>(buf) + buffer_offset, sizeof(buf) - buffer_offset - 1, k_receiver_ack_timeout);
+
+		if (read_result < 0) {
+			SEP_WARN("SBF read error");
+			return PX4_ERROR;
+		}
+
+		// Sanitize the data so it doesn't contain any `0` values.
+		for (size_t i = buffer_offset; i < buffer_offset + read_result; i++) {
+			if (buf[i] == 0) {
+				buf[i] = 1;
+			}
+		}
+
+		buffer_offset += read_result;
+
+		// Make sure the current buffer is a valid string.
+		buf[buffer_offset] = '\0';
+
+		char* port_name_address = strstr(buf, ">");
+
+		// Check if we found a port candidate.
+		if (buffer_offset > 4 && port_name_address != nullptr) {
+			size_t port_name_offset = reinterpret_cast<size_t>(port_name_address) - reinterpret_cast<size_t>(buf) - 4;
+			for (size_t i = 0; i < 4; i++) {
+				port_name[i] = buf[port_name_offset + i];
+			}
+			// NOTE: This limits the ports to serial and USB ports only. Otherwise the detection doesn't work correctly.
+			if (strstr(port_name, "COM") != nullptr || strstr(port_name, "USB") != nullptr) {
+				response_detected = true;
+				break;
+			}
+		}
+
+		if (buffer_offset + 1 >= sizeof(buf)) {
+			// Copy the last 3 bytes such that a half port isn't lost.
+			for (int i = 0; i < 4; i++) {
+				buf[i] = buf[sizeof(buf) - 4 + i];
+			}
+			buffer_offset = 3;
+		}
+	} while (timeout_time > hrt_absolute_time());
+
+	if (!response_detected) {
+		SEP_WARN("No valid serial port detected");
+		return PX4_ERROR;
+	} else {
+		SEP_INFO("Serial port found: %s", port_name);
+		return PX4_OK;
+	}
+}
+
+int SeptentrioDriver::configure()
+{
+	char msg[k_max_command_size] {};
+
+	// Passively detect receiver baud rate.
+	uint32_t detected_receiver_baud_rate = detect_receiver_baud_rate(true);
+
+	if (detected_receiver_baud_rate == 0) {
+		SEP_INFO("CONFIG: failed baud detection");
+		return PX4_ERROR;
+	}
+
+	// Set same baud rate on our end.
+	if (set_baudrate(detected_receiver_baud_rate) != PX4_OK) {
+		SEP_INFO("CONFIG: failed local baud rate setting");
+		return PX4_ERROR;
+	}
+
+	char com_port[5] {};
+
+	// Passively detect receiver port.
+	if (detect_serial_port(com_port) != PX4_OK) {
+		SEP_INFO("CONFIG: failed port detection");
+		return PX4_ERROR;
+	}
+
+	// We should definitely match baud rates and detect used port, but don't do other configuration if not requested.
+	// This will force input on the receiver. That shouldn't be a problem as it's on our own connection.
+	if (!_automatic_configuration) {
```

Comment:
This needs to be after the baud rate detection.

Reply:
Do you mean right after line 742 with the `detect_receiver_baud_rate()`? Currently it's after baud rate detection, setting baud rate on flight controller and port detection as none of them touch the state of the receiver (only sends `getEchoMessage` and force input with `SSSSSSSSSS` which affects only the attached port). I've done it this way so there can be no confusion in the future about why sending commands in other parts of the driver doesn't work, as the baud rates wouldn't be matched. Automatic configuration doesn't mean that users couldn't send commands using the MAVLink console in a future use case for example. Also the current `septentrio reset <hot|warm|cold>` commands require matched baud rates already.